runtime/src/enclave_rpc/demux.rs

```diff
@@ -57,164 +48,293 @@ impl From<Error> for crate::types::Error {
+    inner: SharedSession,
+}
+
+impl SessionMeta {
+    /// Key for ordering in the by-idle-time index.
+    fn by_time_key(&self) -> SessionByTimeKey {
+        (self.last_access_time, self.peer_id.clone(), self.session_id)
+    }
+}
+
+/// Session indices and management operations.
+struct Sessions {
+    /// Session builder.
+    builder: Builder,
+    /// Maximum number of sessions.
+    max_sessions: usize,
+    /// Maximum number of sessions per peer.
+    max_sessions_per_peer: usize,
+    /// Stale session timeout (in seconds).
+    stale_session_timeout: i64,
+
+    /// A map of sessions for each peer.
+    by_peer: HashMap<PeerID, HashMap<SessionID, SessionMeta>>,
+    /// A set of all sessions, ordered by idle time.
+    by_idle_time: BTreeSet<SessionByTimeKey>,
+}
+
+impl Sessions {
+    /// Create a new session management instance.
+    fn new(
+        builder: Builder,
+        max_sessions: usize,
+        max_sessions_per_peer: usize,
+        stale_session_timeout: i64,
+    ) -> Self {
+        Self {
+            builder,
+            max_sessions,
+            max_sessions_per_peer,
+            stale_session_timeout,
+            by_peer: HashMap::new(),
+            by_idle_time: BTreeSet::new(),
+        }
+    }
+
+    /// Create a new multiplexed session.
+    fn create_session(
+        builder: Builder,
+        peer_id: PeerID,
+        session_id: SessionID,
+        now: i64,
+    ) -> SessionMeta {
+        SessionMeta {
+            inner: Arc::new(tokio::sync::Mutex::new(MultiplexedSession {
+                peer_id: peer_id.clone(),
+                session_id,
+                inner: builder.build_responder(),
+            })),
+            peer_id,
+            session_id,
+            last_access_time: now,
+        }
+    }
+
+    /// Fetch an existing session given its identifier or create a new one.
+    fn get_or_create(
+        &mut self,
+        peer_id: PeerID,
+        session_id: SessionID,
+    ) -> Result<(SharedSession, bool), Error> {
+        let now = insecure_posix_time();
+
+        // Check if peer exists.
+        if let Some(sessions) = self.by_peer.get_mut(&peer_id) {
+            // Check if the session exists. If so, return it.
+            if let Some(session) = sessions.get_mut(&session_id) {
+                // Remove old idle time.
+                self.by_idle_time.remove(&session.by_time_key());
+                // Update idle time.
+                session.last_access_time = insecure_posix_time();
+                self.by_idle_time.insert(session.by_time_key());
+
+                return Ok((session.inner.clone(), false));
+            }
+
+            // Check if the peer has max sessions or if no more sessions are available globaly. If
+            // so, remove the oldest or return an error.
+            if sessions.len() >= self.max_sessions_per_peer
+                || self.by_idle_time.len() >= self.max_sessions
```

Comment:
Why? I think having this here makes more sense, e.g. imagine a peer has two out of four sessions open, but there are no more sessions globally. Without this a new connection would fail instead of evicting the other peer's session? Which I don't think is optimal?

Reply:
No. If peer has 2/4, this `if` should be skipped, and the next `if` should take care to remove one global session.

Reply:
But in that case there may be no sessions that are stale enough? While here we force remove the oldest one.

Reply:
Not sure what is better. You can keep as it is.